dbms/src/Flash/Planner/plans/PhysicalProjection.cpp

```diff
@@ -38,20 +38,12 @@ PhysicalPlanNodePtr PhysicalProjection::build(
     ExpressionActionsPtr project_actions = PhysicalPlanHelper::newActions(child->getSampleBlock(), context);
 
     NamesAndTypes schema;
-    NamesWithAliases project_aliases;
-    UniqueNameGenerator unique_name_generator;
     for (const auto & expr : projection.exprs())
     {
         auto expr_name = analyzer.getActions(expr, project_actions);
         const auto & col = project_actions->getSampleBlock().getByName(expr_name);
-
-        String alias = unique_name_generator.toUniqueName(col.name);
-        project_aliases.emplace_back(col.name, alias);
-        schema.emplace_back(alias, col.type);
+        schema.emplace_back(col.name, col.type);
     }
-    /// TODO When there is no alias, there is no need to add the project action.
-    /// https://github.com/pingcap/tiflash/issues/3921
-    project_actions->add(ExpressionAction::project(project_aliases));
```

Comment:
So the basic idea is `PhysicalProjection` itself does not add `project` action to prune column, and leave it to the finalize stage to remove useless column?

Reply:
yes.